### ISSUE TITLE
bowtie2-build runs without error but without constructing .rev files

### ISSUE THREAD

**Issue by AnnaSOFI** [BenLangmead/bowtie2#194]
Hi,

I have just started learning bowtie2. I encountered a problem with database indexing with bowtie2-build command: indexing runs through but the output doesn't have .rev files, thus preventing me from doing alignments to the particular database.
The db file to be indexed is a FASTA file, that is a collection of bacterial genomes.
Previously I have successfully indexed and aligned to a db of the same kind (merged fasta files), so now I have no clue what is wrong with the setting or what is going on. I get the 4 other files in reasonable time, although multiple memory-usage test were performed by the software before passing.

I could not find any help online regarding this question.
I wonder what I'm doing wrong?

BRs,

Anna

**Comment by ch4rr0** [comment]
Would it be possible to share the FASTA?

**Comment by AnnaSOFI** [comment]
Hi, 

The FASTA file is too big to post here, but it has been downloaded from here:

https://www.hmpdacc.org/hmp/catalog/grid.php?dataset=genomic&hmp_isolation_body_site=gastrointestinal_tract

Anna

**Comment by whidbeyc** [comment]
Hi,
I'm having a very similar problem trying to run bowtie2-build on two different FASTAs (one from metaphlan2's marker database and another that was built from chocophlan). Was there ever a solution found to this? I'm not actually getting an error message, but I think its stopping before it actually finishing the index (see below for the output). This is on version 2.3.4.1, I used bioconda to install.
```
INFO: Command: bowtie2-build-s --wrapper basic-0 -f Humann2Test/A1HF_humann2_temp/A1HF_custom_chocophlan_database.ffn Humann2Test/A1HF_humann2_temp/A1HF_bowtie2_index
Settings:
  Output files: "Humann2Test/A1HF_humann2_temp/A1HF_bowtie2_index.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  Humann2Test/A1HF_humann2_temp/A1HF_custom_chocophlan_database.ffn
Building a SMALL index
Reading reference sizes
  Time reading reference sizes: 00:00:00
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:00
bmax according to bmaxDivN setting: 6990730
Using parameters --bmax 5243048 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 5243048 --dcv 1024
Constructing suffix-array element generator
```
If a solution or workaround was found, I'd be very interested to know.
Thanks for your help,
Chris

**Comment by AnnaSOFI** [comment]
Hi,

I did not get official advice on this one, but now it is working.
I split my enormous FASTA file into two (8.4GB --> 4.2GB) just yesterday and the indexing seemed now to be working for the first part at least (I did not yet run the alignment but at least the .rev files have appeared after indexing ins complete)!
Could you try this approach also? ☺
I think I’m going to align my data against both of the FASTA files and then combine the alignment sam files.

BRs,

Anna Sorjamaa


Anna Sorjamaa, PhD student
Research Group of Docent Justus Reunanen
Biocenter Oulu / Cancer Research and Translational Medicine Research Unit
University of Oulu
Aapistie 5, P.O. box 5281
90014 University of Oulu, Finland
Tel: +35845 879 0354
Email: anna.sorjamaa@oulu.fi<mailto:anna.sorjamaa@oulu.fi>

**Comment by lcstewart** [comment]
Hi,
I'm having exactly this problem as well - I installed bowtie2 automatically as part of humann2 (version 2.3.4.1), and bowtie2-build will not build .rev.bt2 files from any fasta file I give it, including as part of the MetaPhlAn pipeline indexing the database, but doesn't throw any errors either. It also produces a .2.bt2 file of 0kb, and (for the failed MetaPhlAn database index, where I have a comparison) the other files are smaller than the correctly-indexed database files provided by the makers of that software. 

Really keen to figure out if this is a bowtie problem, a server problem, a data problem, or something else!

This is the output I get when I try:

`Settings:
  Output files: "04_MAPPING/contigs.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  03_CONTIGS/contigs.fa
Building a SMALL index
Reading reference sizes
  Time reading reference sizes: 00:00:01
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:01
bmax according to bmaxDivN setting: 17568529
Using parameters --bmax 13176397 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 13176397 --dcv 1024
Constructing suffix-array element generator
`

**Comment by ch4rr0** [comment]
I have already begun looking into this. I will update this thread as soon as I have more information.

**Comment by ch4rr0** [comment]
@whidbeyc -- can you share your command line? How big is your index?
@AnnaSOFI -- I am not able to access the files on the website you provided. Would it be possible to share the link some other way? Maybe a dropbox link?

**Comment by bbsunchen** [comment]
Same problem here.
I install bowtie2 2.3.4.1 with conda.
Build the index of a fungi genome with no errors.
The fungi genome is ~10^7 bp. 
The command I used: 
`bowtie2-build -f fungi.fa fungi_bt2`
`bowtie2 --local -x fungi_bt2 -f -U contigs.fasta -S bowtie.contigs.fungi.sam -p 1`

**Comment by uribe-convers** [comment]
Same here! 
I'm trying to build an index from a fasta file that contains ~360,000 contigs (the file size is 194Mb). I'm using bowtie2 version 2.3.4.1 installed from conda (`conda install --yes -c bioconda bowtie2=2.3.4.1`) with the following command: `bowtie2-build final.contigs.fa contig_index` . 

The program runs with no errors and finishes quickly but generates only four files (contig_index.1.bt2 [5.5Mb], contig_index.2.bt2 [**_0Mb_**], contig_index.3.bt2 [3.1Mb], and contig_index.4.bt2 [45Mb]), but no 
.rev1.bt2 or .rev2.bt2 files.

Here is the output I get when I try to build the indices:
```
Settings:
  Output files: "contig_index.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  /home/ubuntu/output_169_subset/final.contigs.fa
Building a SMALL index
Reading reference sizes
  Time reading reference sizes: 00:00:02
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:01
bmax according to bmaxDivN setting: 46852187
Using parameters --bmax 35139141 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 35139141 --dcv 1024
Constructing suffix-array element generator
```

Any advice?
Thanks,
Simon

**Comment by uribe-convers** [comment]
Quick update for people who need a quick fix, installing from the other conda link works (at least the `bowtie2-build`). Here is the link: `conda install -c bioconda/label/broken bowtie2 `

**Comment by ch4rr0** [comment]
I probably should have done this before jumping into debugging code, but I just tried building an index with a `conda`--installed `bowtie2` build and got the same output. This issue seems to be isolated to the `conda` build of `bowtie2` as evidenced below:

Conda:
```
root@a50a2aa0a1ff:/bowtie2# bowtie2-build example/reference/lambda_virus.fa out
Settings:
  Output files: "out.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  example/reference/lambda_virus.fa
Building a SMALL index
Reading reference sizes
  Time reading reference sizes: 00:00:00
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:00
bmax according to bmaxDivN setting: 12125
Using parameters --bmax 9094 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 9094 --dcv 1024
Constructing suffix-array element generator
root@a50a2aa0a1ff:/bowtie2# ls out*
out.1.bt2  out.2.bt2  out.3.bt2  out.4.bt2
```

Local build:
```
root@a50a2aa0a1ff:/bowtie2# ./bowtie2-build example/reference/lambda_virus.fa out
Settings:
  Output files: "out.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  example/reference/lambda_virus.fa
Building a SMALL index
Reading reference sizes
  Time reading reference sizes: 00:00:00
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:00
bmax according to bmaxDivN setting: 12125
Using parameters --bmax 9094 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 9094 --dcv 1024
Constructing suffix-array element generator
Building DifferenceCoverSample
  Building sPrime
  Building sPrimeOrder
  V-Sorting samples
  V-Sorting samples time: 00:00:00
  Allocating rank array
  Ranking v-sort output
  Ranking v-sort output time: 00:00:00
  Invoking Larsson-Sadakane on ranks
  Invoking Larsson-Sadakane on ranks time: 00:00:00
  Sanity-checking and returning
Building samples
Reserving space for 12 sample suffixes
Generating random suffixes
QSorting 12 sample offsets, eliminating duplicates
QSorting sample offsets, eliminating duplicates time: 00:00:00
Multikey QSorting 12 samples
  (Using difference cover)
  Multikey QSorting samples time: 00:00:00
Calculating bucket sizes
Splitting and merging
  Splitting and merging time: 00:00:00
Avg bucket size: 48502 (target: 9093)
Converting suffix-array elements to index image
Allocating ftab, absorbFtab
Entering Ebwt loop
Getting block 1 of 1
  No samples; assembling all-inclusive block
  Sorting block of length 48502 for bucket 1
  (Using difference cover)
  Sorting block time: 00:00:00
Returning block of 48503 for bucket 1
Exited Ebwt loop
fchr[A]: 0
fchr[C]: 12334
fchr[G]: 23696
fchr[T]: 36516
fchr[$]: 48502
Exiting Ebwt::buildToDisk()
Returning from initFromVector
Wrote 4210730 bytes to primary EBWT file: out.1.bt2
Wrote 12132 bytes to secondary EBWT file: out.2.bt2
Re-opening _in1 and _in2 as input streams
Returning from Ebwt constructor
Headers:
    len: 48502
    bwtLen: 48503
    sz: 12126
    bwtSz: 12126
    lineRate: 6
    offRate: 4
    offMask: 0xfffffff0
    ftabChars: 10
    eftabLen: 20
    eftabSz: 80
    ftabLen: 1048577
    ftabSz: 4194308
    offsLen: 3032
    offsSz: 12128
    lineSz: 64
    sideSz: 64
    sideBwtSz: 48
    sideBwtLen: 192
    numSides: 253
    numLines: 253
    ebwtTotLen: 16192
    ebwtTotSz: 16192
    color: 0
    reverse: 0
Total time for call to driver() for forward index: 00:00:00
Reading reference sizes
  Time reading reference sizes: 00:00:00
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:00
  Time to reverse reference sequence: 00:00:00
bmax according to bmaxDivN setting: 12125
Using parameters --bmax 9094 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 9094 --dcv 1024
Constructing suffix-array element generator
Building DifferenceCoverSample
  Building sPrime
  Building sPrimeOrder
  V-Sorting samples
  V-Sorting samples time: 00:00:00
  Allocating rank array
  Ranking v-sort output
  Ranking v-sort output time: 00:00:00
  Invoking Larsson-Sadakane on ranks
  Invoking Larsson-Sadakane on ranks time: 00:00:00
  Sanity-checking and returning
Building samples
Reserving space for 12 sample suffixes
Generating random suffixes
QSorting 12 sample offsets, eliminating duplicates
QSorting sample offsets, eliminating duplicates time: 00:00:00
Multikey QSorting 12 samples
  (Using difference cover)
  Multikey QSorting samples time: 00:00:00
Calculating bucket sizes
Splitting and merging
  Splitting and merging time: 00:00:00
Avg bucket size: 48502 (target: 9093)
Converting suffix-array elements to index image
Allocating ftab, absorbFtab
Entering Ebwt loop
Getting block 1 of 1
  No samples; assembling all-inclusive block
  Sorting block of length 48502 for bucket 1
  (Using difference cover)
  Sorting block time: 00:00:00
Returning block of 48503 for bucket 1
Exited Ebwt loop
fchr[A]: 0
fchr[C]: 12334
fchr[G]: 23696
fchr[T]: 36516
fchr[$]: 48502
Exiting Ebwt::buildToDisk()
Returning from initFromVector
Wrote 4210730 bytes to primary EBWT file: out.rev.1.bt2
Wrote 12132 bytes to secondary EBWT file: out.rev.2.bt2
Re-opening _in1 and _in2 as input streams
Returning from Ebwt constructor
Headers:
    len: 48502
    bwtLen: 48503
    sz: 12126
    bwtSz: 12126
    lineRate: 6
    offRate: 4
    offMask: 0xfffffff0
    ftabChars: 10
    eftabLen: 20
    eftabSz: 80
    ftabLen: 1048577
    ftabSz: 4194308
    offsLen: 3032
    offsSz: 12128
    lineSz: 64
    sideSz: 64
    sideBwtSz: 48
    sideBwtLen: 192
    numSides: 253
    numLines: 253
    ebwtTotLen: 16192
    ebwtTotSz: 16192
    color: 0
    reverse: 1
Total time for backward call to driver() for mirror index: 00:00:00
root@a50a2aa0a1ff:/bowtie2# ls out*
out.1.bt2  out.2.bt2  out.3.bt2  out.4.bt2  out.rev.1.bt2  out.rev.2.bt2  outq.cpp  outq.h
```

**Comment by ch4rr0** [comment]
We do provide pre-built `bowtie2` packages. Any objections to trying those?

**Comment by uribe-convers** [comment]
I’m using bowtie on an instance on AWS and I’ve had problems with pre-built software before. Any chance of fixing the conda built or should I get the source code and try it to build it myself?

Thanks for looking into this!

—
<Sent from a mobile device, please excuse any typo>

**Comment by ch4rr0** [comment]
I'll see what I can do with regards to the conda build tomorrow. I will keep you posted.

**Comment by lcstewart** [comment]
Yeah, my IT department won't grant permissions for installing software myself on our bioinformatics server except through conda or pip, so an updated/working conda install would be very handy.

**Comment by ch4rr0** [comment]
I put together a script that works within the conda framework to replace the problematic `bowtie2` binaries with ones that work. Here's what it does:

* Uses `conda-build` to build `bowtie2` locally (with `tbb` and `libz` statically linked) then creates a local `conda` package
* Deletes your existing `bowtie2` runtime and uses the `conda install` command to replace it with the local build

You can find out more about this process [here](https://conda.io/docs/user-guide/tutorials/build-pkgs.html).

Once the package has been build, `conda install` will output the following

```
The following NEW packages will be INSTALLED:

    bowtie2: 2.3.4.1-py27h6bb024c_1 local

Proceed ([y]/n)? y
```

**Make sure that the package is to be installed is marked `local` and not `bioconda` as seen above**

Copy the below script verbatim and save to file e.g. `bt2_conda.sh`. From the command line execute the script by issuing the command: `bash bt2_conda.sh`. Let me know if you encounter any issues.

```
#!/bin/bash

if [[ ! -e `which conda-build` ]]; then
    conda install conda-build
fi

if [[ ! -e `which wget` && ! -e `which curl` ]]; then
    echo "Please make sure that either curl or wget is installed"
    exit 1
fi

function cleanup {
    files_to_remove=`ls /tmp/bowtie2`
    for f in $files_to_remove; do
        echo "Deleting $f"
        rm -f $f
    done
    rm -r /tmp/bowtie2
}

trap cleanup EXIT

mkdir /tmp/bowtie2 && cd /tmp/bowtie2
cat <<EOF > build.sh
#!/bin/bash
LDFLAGS=""
make static-libs && make RELEASE_BUILD=1

binaries="\
bowtie2 \
bowtie2-align-l \
bowtie2-align-s \
bowtie2-build \
bowtie2-build-l \
bowtie2-build-s \
bowtie2-inspect \
bowtie2-inspect-l \
bowtie2-inspect-s \
"
directories="scripts"
pythonfiles="bowtie2-build bowtie2-inspect"

PY3_BUILD="\${PY_VER%.*}"

if [ \$PY3_BUILD -eq 3 ]; then
    for i in \$pythonfiles; do
        2to3 --write \$i
    done
fi

for i in \$binaries; do
    cp \$i \$PREFIX/bin && chmod +x \$PREFIX/bin/\$i
done

for d in \$directories; do
    cp -r \$d \$PREFIX/bin
done
EOF

cat <<EOF > meta.yaml
{% set version = "2.3.4.1" %}


package:
  name: bowtie2
  version: {{ version }}

source:
  url: http://downloads.sourceforge.net/project/bowtie-bio/bowtie2/{{ version }}/bowtie2-{{ version }}-source.zip
  sha256: a1efef603b91ecc11cfdb822087ae00ecf2dd922e03c85eea1ed7f8230c119dc
  patches:
    - bowtie2.patch

build:
  number: 1

requirements:
  build:
    - {{ compiler('cxx') }}
  host:
    - python
  run:
    - python
    - perl

test:
  commands:
    - bowtie2 --help
    - bowtie2-align-l --help
    - bowtie2-align-s --help
    - bowtie2-build --help
    - bowtie2-build-l --help
    - bowtie2-build-s --help
    - bowtie2-inspect --help
    - bowtie2-inspect-l --help
    - bowtie2-inspect-s --help

about:
  home: 'http://bowtie-bio.sourceforge.net/bowtie2/index.shtml'
  license: GPLv3
  summary: Fast and sensitive read alignment

extra:
  identifiers:
    - biotools:bowtie2
    - doi:10.1038/nmeth.1923
EOF

cat <<EOF > bowtie2.patch
--- bowtie2.orig        2017-01-20 19:41:25.706765000 -0500
+++ bowtie2     2017-01-20 16:23:38.574188000 -0500
@@ -38,10 +38,10 @@
  my (\$vol,\$script_path,\$prog);
  \$prog = File::Spec->rel2abs( __FILE__ );

-while (-f \$prog && -l \$prog){
-    my (undef, \$dir, undef) = File::Spec->splitpath(\$prog);
-    \$prog = File::Spec->rel2abs(readlink(\$prog), \$dir);
-}
+#while (-f \$prog && -l \$prog){
+#    my (undef, \$dir, undef) = File::Spec->splitpath(\$prog);
+#    \$prog = File::Spec->rel2abs(readlink(\$prog), \$dir);
+#}

  (\$vol,\$script_path,\$prog)
                  = File::Spec->splitpath(\$prog);
EOF

cd -

conda-build /tmp/bowtie2

if [[ $? -ne 0 ]]; then
    echo "Build failed... exiting"
    exit 1
fi

echo "Build complete... Uninstalling your current bowtie2 runtime."
echo "Would you like to continue? yes/no?"
read ans
case "$ans" in
    yes)
        conda uninstall bowtie2
        ;;
    *)
        echo Exiting...
        exit 1
        ;;
esac
conda install --use-local bowtie2
```

**Comment by uribe-convers** [comment]
Thank you so much! 
I’ll try it tomorrow morning and report back if I have any problems! 
Simon

**Comment by microsat2018** [comment]
I encountered the same problem. The bowtie2-build is 2.3.4.1 from conda.
I actually used a very small fasta file
>a1
CATGTCCAGCTTCTCTTCAGTACCGCTCACCAGCCTAGGTGGGACCACTGACTGTGAGTC
TGCAGTGGCCACCGCCCAGTCTCTGTGTCTCAAGCTCCAAGAGACGGTCACACACTAACC
TGCAAGCCAAGGCTGGTGACTTTGACCATCCCTAACGCATGAGTTTTCCATGGAAACCTG
GTCGGTGAACCTGACACGAAATTCCCAATTCCCCTTTACTCTGTACTGTGTGGCTGGTGC
TCTTGTTTTCGTTCTCTCTCTCTCTCTCTCTCTCTCTCAAGTTGATTCCTCCATGTTGCT
TTACAGAGACCTGCCAACTACCCAGGAATGTAAAAGCATTCATAGTATTTGTCTAGTAGA

No the two .rev.*.bt2 files
Can anyone help with this issue?

**Comment by uribe-convers** [comment]
I ended up being able to use the pre-built `bowtie2` binary and works well. Get it from here: `https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.3.4.1`

or from the command line
```
wget https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.3.4.1/bowtie2-2.3.4.1-linux-x86_64.zip/download -O bowtie2-2.3.4.1-linux-x86_64.zip
```

**Comment by nick-youngblut** [comment]
My research group is having the same (or similar) problem when running `bioconda::bowtie2=2.3.4.1` as part of `bioconda::humann2=0.11.1`. The error always something like:

```
Building a SMALL index
Index is corrupt: File size for /tmp/global/LLMGP_82144987094/MI-208-H/MI-208-H_humann2_temp/MI-208-H_bowtie2_index.1.bt2 should have been 132894396 but is actually 0.
Index is corrupt: File size for /tmp/global/LLMGP_82144987094/MI-208-H/MI-208-H_humann2_temp/MI-208-H_bowtie2_index.2.bt2 should have been 63736900 but is actually 0.
Please check if there is a problem with the disk or if disk is full.
Error: Encountered internal Bowtie 2 exception (#1)
```

...and sometimes in the "rev" index files are are zero-sized. 

This error seems to happen stochastically for us, and usually only when the I/O load is really heavy. If I rerun the "problematic" samples in isolation (instead of 20-30 parallel humann2 jobs), then they usually work. Maybe it's some sort of file latency issue?

**Comment by ch4rr0** [comment]
I recently had a pull request merged that changes the way conda builds bowtie 2. This new build process will be in effect for our most recent release of bowtie 2, v2.3.4.2. The updated build process is identical to the way we build our bowtie 2 binaries for distribution. The resulting binaries will have all dependencies statically linked which should solve what I think has been the root of this issue, the dynamically linked TBB library. Please give this new version a try and let me know if this problem still persists.

**Comment by ch4rr0** [comment]
I will be closing this thread. Feel free to reopen if you believe that v2.3.4.2 has not addressed this issue.

**Comment by domenico-simone** [comment]
Hello,

I am re-opening this thread since I'm not able to get proper index files with either 2.3.3.1 or 2.3.4.1, but I get them with 2.2.9. I am working on a cluster (not through conda). If it matters, the input files contains nearly 1M sequences. The stdout is 

```
[domeni@r102 new_contig_mappings]$ bowtie2-build thawponds_assembly.fa thawponds_assembly
Settings:
  Output files: "thawponds_assembly.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  thawponds_assembly.fa
Building a SMALL index
Reading reference sizes
  Time reading reference sizes: 00:00:18
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:10
bmax according to bmaxDivN setting: 569603352
Using parameters --bmax 427202514 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 427202514 --dcv 1024
Constructing suffix-array element generator
Building DifferenceCoverSample
  Building sPrime
  Building sPrimeOrder
  V-Sorting samples
  V-Sorting samples time: 00:00:48
  Allocating rank array
  Ranking v-sort output
  Ranking v-sort output time: 00:00:15
  Invoking Larsson-Sadakane on ranks
  Invoking Larsson-Sadakane on ranks time: 00:00:28
  Sanity-checking and returning
Building samples
Reserving space for 12 sample suffixes
Generating random suffixes
QSorting 12 sample offsets, eliminating duplicates
QSorting sample offsets, eliminating duplicates time: 00:00:00
Multikey QSorting 12 samples
  (Using difference cover)
  Multikey QSorting samples time: 00:00:00
Calculating bucket sizes
Splitting and merging
  Splitting and merging time: 00:00:00
Avg bucket size: 2.27841e+09 (target: 427202513)
Converting suffix-array elements to index image
Allocating ftab, absorbFtab
Entering Ebwt loop
Getting block 1 of 1
  No samples; assembling all-inclusive block
```

and I get all files but the rev ones. I've read the previous messages in this thread, it's not clear to me if I should ask the admins to re-install the previous 2.3.x versions or install 2.3.4.2.

Many thanks,

Domenico

**Comment by ch4rr0** [comment]
Hello, 

How big is your input file? Are you using one of our pre-built binary packages?

**Comment by MarieLataretu** [comment]
Hi,

I still have the problem, that bowtie2-build does not create rev files and/or produces segmentation faults.
Here are the versions I tested:
- version 2.2.9 (conda create -n bowtie2-broken -c bioconda/label/broken bowtie2)
    - segmentation fault, no rev files
- version 2.3.5 (conda create -n bowtie2 -c bioconda bowtie2)
    - segmentation fault, no rev files
- version 2.2.9 (bowtie2-2.2.9-linux-x86_64.zip)
    - segmentation fault, no rev files
- version 2.3.5 (bowtie2-2.3.5.1-linux-x86_64.zip)
    - no segmentation fault, no rev files

**Comment by ch4rr0** [comment]
Hello @MarieLataretu ,

Can you please share the FASTA and the command line for the index that you are trying to build? This issue seems to happen sporadically and hence has been difficult to debug.

**Comment by MarieLataretu** [comment]
Hi,

this is the input fasta (unzipped):
```
ftp://ftp.ensembl.org/pub/release-92/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.primary_assembly.fa.gz
```

here we add 'chr' to chromosomes as prefix:
```
sed -r '/^>/ s/>([1-9MXY])/>chr\1/' Mus_musculus.GRCm38.dna.primary_assembly.fa > Mus_musculus.GRCm38.dna.primary_assembly.chr.fa
```

and this is the command:
```
mkdir bowtie2-index
nice [/path/to/]bowtie2-build -t 20 /path/to/genome/Mus_musculus.GRCm38.dna.primary_assembly.chr.fa bowtie2-index/Mus_musculus.GRCm38.dna.primary_assembly.chr &> bowtie2-index/bowtie2-build.log
```

I have the same problem without the prefix and when the fasta file is in the same directory as the output.

**Comment by ch4rr0** [comment]
If you have the `bowtie2-build-s-debug` binary available can you try using it when building the index? I'd also appreciate it if you can save and share the debug output with me.

**Comment by MarieLataretu** [comment]
Sure!

the command is now:
```
nice /path/to/bowtie2-build-s-debug -t 20 /path/to/Mus_musculus.GRCm38.dna.primary_assembly.chr.fa bowtie2-index/Mus_musculus.GRCm38.dna.primary_assembly.chr.debug &> bowtie2-index/debug.log
```

and the resulting log file:
```
Settings:
  Output files: "bowtie2-index/Mus_musculus.GRCm38.dna.primary_assembly.debug.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 20
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: enabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  /data/fass1/genomes/Eukaryots/mus_musculus_done/03052018/Mus_musculus.GRCm38.dna.primary_assembly.fa
Building a SMALL index
Reading reference sizes
  Time reading reference sizes: 00:00:47
assert_leq: expected (20) <= (16)
bt2_idx.h:208
bowtie2-build-s-debug: bt2_idx.h:208: bool EbwtParams::repOk() const: Assertion `0' failed.
```

Only 3 and 4 index files were created (before only the rev files were missing); the running time dropped from > 10 minutes to ~ 1 minute.

Thanks for looking into this!

**Comment by ch4rr0** [comment]
I switched to a dedicated Linux box and now am finally able to recreate the issue consistently. Will update this thread if/when I am able to find the cause of this issue. Thank you for your patience.

**Comment by ch4rr0** [comment]
Did you pass in `-t 20` with the intention of increasing the number of threads? When run in debug mode Bowtie2 will fail on the assertion that ftabChars should be less than or equal to 16, but you passed in the value 20. This also seems to affect the release build as well preventing Bowtie 2 from creating the reverse indexes. Try removing `-t 20`, as I did, and see if the reverse files get created.

**Comment by alesanzdro** [comment]
Apparently I had the same error, (Bowtie 2 version 2.3.5) "-rev" files were not generated. And it seems that adding the "-c" option has been fixed.

**conda install -c bioconda bowtie2** 
**bowtie2-build -c --threads 24 $genome $out  > out.msg 2> out.err**

At the moment it is running well and in parallel.

**Comment by MarieLataretu** [comment]
Aha,
`-t 20 --thread 20` does not work, whereas `--thread 20` works, as well as no `-t` and `--threads` parameters set!

The version is bowtie2-2.3.5.1.

**Comment by YiJessePi** [comment]
I'm having a similar issue. After using conda I've used pre-built bowtie2 (23.5.1) downloaded from: https://netix.dl.sourceforge.net/project/bowtie-bio/bowtie2/2.3.5.1/bowtie2-2.3.5.1-linux-x86_64.zip 
unzipped and executed the build command:
bowtie2-build  ref_db.fa ref_db_packed --threads 2 --large-index --packed
The process ends suddenly , without any out of memory or or core-dump message. 
`Settings:
  Output files: "ref_db_packed.*.bt2l"
  Line rate: 7 (line is 128 bytes)
  Lines per side: 1 (side is 128 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: packed
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 8
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  ref_db.fa
Building a LARGE index
Reading reference sizes
  Time reading reference sizes: 00:16:12
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:11:35
bmax according to bmaxDivN setting: 17192161509
Using parameters --bmax 12894121132 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 12894121132 --dcv 1024 --packed
Constructing suffix-array element generator
Building DifferenceCoverSample
  Building sPrime
  Building sPrimeOrder
  V-Sorting samples
  V-Sorting samples time: 01:59:07
  Allocating rank array
  Ranking v-sort output
  Ranking v-sort output time: 00:58:21
  Invoking Larsson-Sadakane on ranks
  Invoking Larsson-Sadakane on ranks time: 01:26:11
  Sanity-checking and returning
Building samples
Reserving space for 22 sample suffixes
Generating random suffixes
QSorting 22 sample offsets, eliminating duplicates
QSorting sample offsets, eliminating duplicates time: 00:00:00
Multikey QSorting 22 samples
  (Using difference cover)
  Multikey QSorting samples time: 00:00:00
Calculating bucket sizes
Splitting and merging
  Splitting and merging time: 00:00:00
Split 4, merged 12; iterating...
Splitting and merging
  Splitting and merging time: 00:00:00
Split 3, merged 1; iterating...
Splitting and merging
  Splitting and merging time: 00:00:00
Avg bucket size: 9.82409e+09 (target: 12894121131)
Converting suffix-array elements to index image
Allocating ftab, absorbFtab
Entering Ebwt loop
Getting block 1 of 14
Getting block 2 of 14
  Reserving size (12894121132) for bucket 1
  Reserving size (12894121132) for bucket 2
  Calculating Z arrays for bucket 1
  Calculating Z arrays for bucket 2
  Entering block accumulator loop for bucket 1:
  Entering block accumulator loop for bucket 2:
  bucket 1: 10%
  bucket 2: 10%
  bucket 1: 20%
  bucket 2: 20%
  bucket 1: 30%
  bucket 2: 30%
  bucket 1: 40%
  bucket 2: 40%
  bucket 1: 50%
  bucket 2: 50%
  bucket 1: 60%
  bucket 2: 60%
  bucket 1: 70%
  bucket 2: 70%
  bucket 1: 80%
  bucket 1: 90%
  bucket 2: 80%
  bucket 1: 100%
  Sorting block of length 12843066619 for bucket 1
  (Using difference cover)
  bucket 2: 90%
  bucket 2: 100%
  Sorting block of length 7007973381 for bucket 2
  (Using difference cover)
  Sorting block time: 06:08:46
Returning block of 7007973382 for bucket 2
Getting block 3 of 14
  Reserving size (12894121132) for bucket 3
  Calculating Z arrays for bucket 3
  Entering block accumulator loop for bucket 3:
  bucket 3: 10%
  bucket 3: 20%
  bucket 3: 30%
  bucket 3: 40%
  bucket 3: 50%
  bucket 3: 60%
  bucket 3: 70%
`
The genome is pretty big (~100GB) so it might be the issue, but I'm not sure since while tracking the usage (both CPU and RAM) it seems to be OK.
Any idea what happening here?
(I think it is not a version issue because I tried both 2.4.3 via conda and pre-built 2.3.5.1).
Thanks!

**Comment by ch4rr0** [comment]
I’m going to publish a beta build today that I think solves this issue. If you can help test the build when it is out that will be appreciated.

**Comment by ch4rr0** [comment]
Having some trouble creating the build. Will try again tomorrow.

**Comment by YiJessePi** [comment]
Thanks. I'll try when it will be ready. 
Just want to update that it do seems to be (also?) a memory issue. I've logged the memory while running and it seemed to  use 100% of memory when it stopped.
Is there anything to do except of using --packed and increasing --offset to reduce RAM usage while building the index?
Thanks again!

**Comment by ch4rr0** [comment]
I don't think there's any other way to decrease the size of index files other than using `--packed`. You can try using more threads to speed up the process. Your issue seems slightly different from what's being reported here. It seems like `bowtie2-build` is stopping prematurely as opposed to running to completion and not generating the reverse index. In any event here is a [link](https://send.firefox.com/download/bb393710962f2758/#0049V_Av4Pr2WitIMKuZFQ) to the build I created earlier. It's a one-off so I didn't bother uploading to GitHub.

Here's the SHA256 sum incase you want to verify the file's integrity (ignore the version number as I didn't bother updating the `VERSION` file in the repo).
```
03d99c3c8b12219422575772b08076cd27c81eff0b4e3f0522fc998456898412  bowtie2-2.3.5.1-linux-x86_64.zip
```

**Comment by YiJessePi** [comment]
Thanks! I'll try it once the current index build will finish. It somehow takes two and a half days...
It enters the Ebwt loop again and again already 12 times (here is one of the twelve):

Entering Ebwt loop
Getting block 1 of 1
  No samples; assembling all-inclusive block
  Ran out of memory; automatically trying more memory-economical parameters.
Using parameters --bmax 1089169922 --dcv 4096
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 1089169922 --dcv 4096 --packed
Constructing suffix-array element generator
Building DifferenceCoverSample
  Building sPrime
  Building sPrimeOrder
  V-Sorting samples
  V-Sorting samples time: 02:28:08
  Allocating rank array
  Ranking v-sort output
  Ranking v-sort output time: 00:35:30
  Invoking Larsson-Sadakane on ranks
  Invoking Larsson-Sadakane on ranks time: 00:35:52
  Sanity-checking and returning
Building samples
Reserving space for 254 sample suffixes
Generating random suffixes
QSorting 254 sample offsets, eliminating duplicates
QSorting sample offsets, eliminating duplicates time: 00:00:00
Multikey QSorting 254 samples
  (Using difference cover)
  Multikey QSorting samples time: 00:00:00
Calculating bucket sizes
Splitting and merging
  Splitting and merging time: 00:00:00
Avg bucket size: 1.37537e+11 (target: 1089169921)
Converting suffix-array elements to index image
Allocating ftab, absorbFtab
Entering Ebwt loop
Getting block 1 of 1
  No samples; assembling all-inclusive block
  Ran out of memory; automatically trying more memory-economical parameters.
Using parameters --bmax 816877442 --dcv 4096
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 816877442 --dcv 4096 --packed
Constructing suffix-array element generator
Building DifferenceCoverSample
  Building sPrime
  Building sPrimeOrder
  V-Sorting samples
  V-Sorting samples time: 02:33:43
  Allocating rank array
  Ranking v-sort output
  Ranking v-sort output time: 00:38:27
  Invoking Larsson-Sadakane on ranks
  Invoking Larsson-Sadakane on ranks time: 00:35:40
  Sanity-checking and returning
Building samples
Reserving space for 338 sample suffixes
Generating random suffixes
QSorting 338 sample offsets, eliminating duplicates
QSorting sample offsets, eliminating duplicates time: 00:00:00
Multikey QSorting 338 samples
  (Using difference cover)
  Multikey QSorting samples time: 00:00:00
Calculating bucket sizes
Splitting and merging
  Splitting and merging time: 00:00:00
Avg bucket size: 1.37537e+11 (target: 816877441)
Converting suffix-array elements to index image
Allocating ftab, absorbFtab

Does it seems to be stucked in infinite loop or it seems normal to you?
My best!

**Comment by MarieLataretu** [comment]
Hi ch4rr0,

- `-threads 20` works fine 
- `-t 20` end with the following message:

```
-t/--ftabChars arg must not exceed 16
Error: Encountered internal Bowtie 2 exception (#1)
```

Command:
```
nice /path/to/bowtie2-2.4.0-beta-linux-x86_64/bowtie2-build-s-debug [-t 20|--threads 20] /path/to/Mus_musculus.GRCm38.dna.primary_assembly.chr.fa bowtie2-index/Mus_musculus.GRCm38.dna.primary_assembly.chr.debug
```

I'll check `-t 15 --threads 20` in the next days.

**Comment by ch4rr0** [comment]
Thanks for the detailed log. I will look into the issue and get back to you once it's been resolved.

**Comment by pmdbio** [comment]
I believe I'm having the same problem.

I'm using MacOS Catalina and zsh. I used bowtie2 pre-compiled binaries.

After trying to use bowtie2-build for the first time it generated 4 files, but no .1.bt2 or .rev.1.bt2.

After trying to run the same command in a different folder, I got this:

`ltelles@ltelles-mac lambda_virus % ~/Desktop/documents_desktop/bioinformatics/bowtie2-2.3.5.1-macos-x86_64/bowtie2-build ~/Desktop/documents_desktop/bioinformatics/bowtie2-2.3.5.1-macos-x86_6/example/reference/lambda_virus.fa lambda_virus
Settings:
  Output files: "lambda_virus.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  /Users/ltelles/Desktop/documents_desktop/bioinformatics/bowtie2-2.3.5.1-macos-x86_6/example/reference/lambda_virus.fa
Error: could not open /Users/ltelles/Desktop/documents_desktop/bioinformatics/bowtie2-2.3.5.1-macos-x86_6/example/reference/lambda_virus.fa
Total time for call to driver() for forward index: 00:00:00
Error: Encountered internal Bowtie 2 exception (#1)
Command: /Users/ltelles/Desktop/documents_desktop/bioinformatics/bowtie2-2.3.5.1-macos-x86_64/bowtie2-build-s --wrapper basic-0 /Users/ltelles/Desktop/documents_desktop/bioinformatics/bowtie2-2.3.5.1-macos-x86_6/example/reference/lambda_virus.fa lambda_virus`

I am sure I used the correct paths for `bowtie2-build` and for the `lambda_virus.fa` files.

**Comment by ch4rr0** [comment]
@MarieLataretu -- I fixed the issue that was causing this in a recent commit, and confirmed it worked by building the GRCm38 without failure. I have not created a build with that includes this commit, but the issue should only affect bowtie2-build when running in debug mode.

@pmdbio -- You appear to have a typo in your path: missing the `4` after in `x86_64`.
~/Desktop/documents_desktop/bioinformatics/bowtie2-2.3.5.1-macos-**x86_6**/example/reference/lambda_virus.fa

@YiJessePi -- For such a large reference, 100GB, it's not unusual for bowtie2-build to take a long time. The output may seem repetitive, but unless bowtie2-build has crashed it's almost guaranteed that it is working on building the index.

**Comment by pmdbio** [comment]
@ch4rr0 Thanks for spotting that. Indeed it was a typo on the path.

**Comment by divyanandu** [comment]
I seem to be having a similar problem. I have so far tried  bowtie2/2.3.4.1 and bowtie2/2.3.0  on our cluster and a pre-built version from https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.3.4.1

In all 3 cases I wasn't able to generate the .rev files for the mouse genome. I was able to generate a .rev file for a very small viral genome. 

I tried to use the bowtie2-build-s-debug command. This was my input
`./bowtie2-2.3.4.1-linux-x86_64/bowtie2-build-s-debug mm10.fa mm10`

(mm10 is the unmasked mouse genome from UCSC)

and got the following output (The run finishes very quickly)
```

Settings:
  Output files: "mm10.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: enabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  mm10.fa
Building a SMALL index
Reading reference sizes
  Time reading reference sizes: 00:01:05
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:29
bmax according to bmaxDivN setting: 663195875
Using parameters --bmax 497396907 --dcv 1024
  Doing ahead-of-time memory usage test
Killed

```

I tried to access the pre-build that was indicated in a previous message but the link doesn't work anymore. Can someone help me?

Thanks in advance!

**Comment by rob123king** [comment]
you just need to set it to produce a large index using the option and not small. it automatically detects and produces a small index sometimes which don't have the rev files but some tools are build to expect a large index

**Comment by joeltorres96** [comment]
@ch4rr0   Hi! I hope to help me pls.
I´m using the subsystem of Linux on windows10 pro, and I dowloaded and run "Bowtie" by UBUNTU, i downloaded UBUNTU from Microsoft Store. My problem is this:
I stay in Ubuntu:

joeltorres@DESKTOP-EBIAPSJ: ~$ bowtie
No index, query, or output file specified!
Usage:
bowtie [options]* {-1 -2 | --12 | --interleaved 
etc...


I dont understand, how can i resolve this? I need your help. I'm new in bioinformatics

**Comment by jgarces02** [comment]
Hi there, I also had this problem and I saw [this comment](https://www.biostars.org/p/233642/#233656) on Biostars... I try to use another reference file without Ns nucleotides (i.e., only those coding for transcripts) and the ".rev" files were (in appearance) correctly generated. I hope that helps!

**Comment by ch4rr0** [comment]
@joeltorres96 — what command are you trying to run? You provided an example of you running bowtie without any parameters which will result in the usage message getting printed.

**Comment by jmwhitha** [comment]
Good afternoon,

I'm getting the error below when I try using bowtie2.4.1.

Output from stderr:
Building a LARGE index
Could not open reference-string index file IMGXXXassemblybowtie2build2.3.bt2l for reading.
This is most likely because your index was built with an older version
(<= 0.9.8.1) of bowtie-build.  Please re-run bowtie-build to generate a new
index (or download one from the Bowtie website) and try again.
Error: Encountered internal Bowtie 2 exception (#1)
Command: bowtie2-align-l --wrapper basic-0 --threads 16 -x IMGXXXassemblybowtie2build2 -S IMGXXXmap2.sam --interleaved clean2.fq
(ERR): bowtie2-align exited with value 1

Output:
IMGXXXassemblybowtie2build2.1.bt2l
IMGXXXassemblybowtie2build2.2.bt2l
IMGXXXassemblybowtie2build2.rev.1.bt2l
IMGXXXassemblybowtie2build2.rev.2.bt2l

So, I don't even get a *.3.bt2l or *4.bt2l output with my build. What's happening?

Thank you, Jason

**Comment by WeichenZhou** [comment]
Hello,

Same problem here, no *.rev.bt2 and the size of *.2.bt2 is zero. I was indexing GRCh38 human genome.

And the same situation as @domenico-simone 
I'm not able to get proper index files with either 2.3.4.3 or 2.4.1, but **I get them with 2.1.0**. I am working on a cluster of our University.

My command line is 
`bowtie2-build GRCh38_canonical_only.fa GRCh38_canonical_only`

The wrong rounds stdout like this:
```
Settings:
  Output files: "GRCh38_canonical_only.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  GRCh38_canonical_only.fa
Building a SMALL index
Reading reference sizes
  Time reading reference sizes: 00:00:25
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:19
bmax according to bmaxDivN setting: 730933138
Using parameters --bmax 548199854 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 548199854 --dcv 1024
Constructing suffix-array element generator
Building DifferenceCoverSample
  Building sPrime
  Building sPrimeOrder
  V-Sorting samples
  V-Sorting samples time: 00:01:14
  Allocating rank array
  Ranking v-sort output
  Ranking v-sort output time: 00:00:18
  Invoking Larsson-Sadakane on ranks
  Invoking Larsson-Sadakane on ranks time: 00:00:39
  Sanity-checking and returning
Building samples
Reserving space for 12 sample suffixes
Generating random suffixes
QSorting 12 sample offsets, eliminating duplicates
QSorting sample offsets, eliminating duplicates time: 00:00:00
Multikey QSorting 12 samples
  (Using difference cover)
  Multikey QSorting samples time: 00:00:00
Calculating bucket sizes
Splitting and merging
  Splitting and merging time: 00:00:00
Avg bucket size: 2.92373e+09 (target: 548199853)
Converting suffix-array elements to index image
Allocating ftab, absorbFtab
Entering Ebwt loop
Getting block 1 of 1
  No samples; assembling all-inclusive block
```

And the successful round goes like this:
```
Settings:
  Output files: "GRCh38_canonical_only.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  GRCh38_canonical_only.fa
Reading reference sizes
  Time reading reference sizes: 00:00:25
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:18
bmax according to bmaxDivN setting: 730933138
Using parameters --bmax 548199854 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 548199854 --dcv 1024
Constructing suffix-array element generator
Building DifferenceCoverSample
  Building sPrime
  Building sPrimeOrder
  V-Sorting samples
  V-Sorting samples time: 00:01:19
  Allocating rank array
  Ranking v-sort output
  Ranking v-sort output time: 00:00:19
  Invoking Larsson-Sadakane on ranks
  Invoking Larsson-Sadakane on ranks time: 00:00:39
  Sanity-checking and returning
Building samples
Reserving space for 12 sample suffixes
Generating random suffixes
QSorting 12 sample offsets, eliminating duplicates
QSorting sample offsets, eliminating duplicates time: 00:00:00
Multikey QSorting 12 samples
  (Using difference cover)
  Multikey QSorting samples time: 00:00:00
Calculating bucket sizes
  Binary sorting into buckets
  10%
  20%
  30%
  40%
  50%
  60%
  70%
  80%
  90%
  100%
  Binary sorting into buckets time: 00:01:41
Splitting and merging
  Splitting and merging time: 00:00:00
Split 1, merged 7; iterating...
  Binary sorting into buckets
  10%
  20%
  30%
  40%
  50%
  60%
  70%
  80%
  90%
  100%
  Binary sorting into buckets time: 00:01:26
Splitting and merging
  Splitting and merging time: 00:00:00
Split 1, merged 0; iterating...
  Binary sorting into buckets
  10%
  20%
  30%
  40%
  50%
  60%
  70%
  80%
  90%
  100%
  Binary sorting into buckets time: 00:01:29
Splitting and merging
  Splitting and merging time: 00:00:00
Avg bucket size: 4.17676e+08 (target: 548199853)
Converting suffix-array elements to index image
Allocating ftab, absorbFtab
Entering Ebwt loop
Getting block 1 of 7
  Reserving size (548199854) for bucket
  Calculating Z arrays
  Calculating Z arrays time: 00:00:00
  Entering block accumulator loop:
  10%
  20%
  30%
  40%
  50%
  60%
  70%
  80%
  90%
  100%
  Block accumulator loop time: 00:00:33
  Sorting block of length 500960528
  (Using difference cover)
...
Returning block of 206598979
Exited Ebwt loop
fchr[A]: 0
fchr[C]: 863155826
fchr[G]: 1459589110
fchr[T]: 2058077499
fchr[$]: 2923732554
Exiting Ebwt::buildToDisk()
Returning from initFromVector
Wrote 978783287 bytes to primary EBWT file: GRCh38_canonical_only.rev.1.bt2
Wrote 730933144 bytes to secondary EBWT file: GRCh38_canonical_only.rev.2.bt2
Re-opening _in1 and _in2 as input streams
Returning from Ebwt constructor
Headers:
    len: 2923732554
    bwtLen: 2923732555
    sz: 730933139
    bwtSz: 730933139
    lineRate: 6
    offRate: 4
    offMask: 0xfffffff0
    ftabChars: 10
    eftabLen: 20
    eftabSz: 80
    ftabLen: 1048577
    ftabSz: 4194308
    offsLen: 182733285
    offsSz: 730933140
    lineSz: 64
    sideSz: 64
    sideBwtSz: 48
    sideBwtLen: 192
    numSides: 15227774
    numLines: 15227774
    ebwtTotLen: 974577536
    ebwtTotSz: 974577536
    color: 0
    reverse: 1
Total time for backward call to driver() for mirror index: 00:52:40
```

**Comment by jmwhitha** [comment]
Good morning,

I realized that I put a "-r" flag which stands for "--no ref" and:
Do not build the NAME.3.bt2 and NAME.4.bt2 portions of the index, which contain a bitpacked version of the reference sequences and are used for paired-end alignment.

...when I intended to put a "-f" flag which stands for "reference file":
The reference input files (specified as <reference_in>) are FASTA files (usually having extension .fa, .mfa, .fna or similar).

That is why I did not get build files ending in 3.bt2l and 4.bt2l.

Thank you,
Jason

**Comment by BJWiley233** [comment]
This just happened to me on Ubuntu 20.04 for version 2.4.1 and 2.4.2 where I used pre-built 2.4.2 as well as building 2.4.2 from source.

**Comment by RxLoutre** [comment]
Hi everyone, just adding my stone to the pile, but I also got this same error very recently.

**Comment by ch4rr0** [comment]
@BJWiley233 — I am leaning towards this being an out of memory issue and if it is I think we should definitely report an error instead of exiting silently. What puzzles me is that some users are saying that they get the reverse index when using v2.2.8/9. I have been comparing the changes made since the aforementioned release and have yet to figure out the cause of this issue.

@RxLoutre — since you’re the most recent victim of this issue would it be possible to provide some details, namely:
* index name and size
* bowtie2-build output
* Output of `ls -l` on the generated index files


Thank you all for your patience and apologies for any inconveniences.

**Comment by ch4rr0** [comment]
Hi all, I pushed a change to the bug fixes branch that hopefully addresses this issue. Going through the comments it seems like most of you experienced this issue when running single threaded. 

If anyone needs a pre-built binary for testing please let me know.

**Comment by apposada** [comment]
Hi, I have found a similar issue when bowtie2-build is called from Trinity (in a conda environment). This bowtie2-build command enters into sleep status, rendering the pipeline on hold as it is waiting for bowtie2-build I/O (bowtie2 version 2.4.2 from bioconda, installed as a trinity dependency). A quick browse through the files at the directory in question reveals that no .rev files have been generated, and the .2.bt2 file is 0B:

```
(trinity_venv) [aperpos@nodo04 chrysalis]$ ls -lha
total 270M
drwxr-xr-x 2 aperpos ska 4,0K jun 14 15:38 .
drwxr-xr-x 4 aperpos ska 4,0K jun 14 15:38 ..
-rw-r--r-- 1 aperpos ska 208M jun 14 15:38 inchworm.DS.fa.min100
-rw-r--r-- 1 aperpos ska  16M jun 14 15:39 inchworm.DS.fa.min100.1.bt2
-rw-r--r-- 1 aperpos ska    0 jun 14 15:38 inchworm.DS.fa.min100.2.bt2
-rw-r--r-- 1 aperpos ska 8,8M jun 14 15:38 inchworm.DS.fa.min100.3.bt2
-rw-r--r-- 1 aperpos ska  38M jun 14 15:38 inchworm.DS.fa.min100.4.bt2
-rw-r--r-- 1 aperpos ska    0 jun 14 15:38 inchworm.DS.fa.min100.ok
(trinity_venv) [aperpos@nodo04 chrysalis]$ 
```

My Trinity run looks like:
`Trinity --seqType fq --max_memory 50G --left $fq1_noribo --right $fq2_noribo --include_supertranscripts --CPU 8 --trimmomatic --verbose --output trinity_denovo_${organism_sample_id} --full_cleanup`

The command that gets "stuck" is:
`/home/aperpos/programs/miniconda3/envs/trinity_venv/bin/bowtie2-build --threads 8 -o 3 /path/to/trinity/working/directory/08_trinity_denovo/trinity_denovo_test/chrysalis/inchworm.DS.fa.min100 /path/to/trinity/working/directory/08_trinity_denovo/trinity_denovo_test/chrysalis/inchworm.DS.fa.min100 1>/dev/null 2>tmp.23580.1623677920.stderr`

The content of tmp.23580.1623677920.stderr is simply:
`Building a SMALL index`

I find two different processes in S status using `htop`:
```

3780    sh -c /home/aperpos/programs/miniconda3/envs/trinity_venv/bin/bowtie2-build --threads 8 -o 3 /path/to/trinity/working/directory/08_trinity_denovo/trinity_denovo_test/chrysalis/inchworm.DS.fa.min100 /path/to/trinity/working/directory/08_trinity_denovo/trinity_denovo_test/chrysalis/inchworm.DS.fa.min100 1>/dev/null 2>tmp.23580.1623677920.stderr
3871    python3 /home/aperpos/programs/miniconda3/envs/trinity_venv/bin/bowtie2-build --threads 8 -o 3 /path/to/trinity/working/directory/08_trinity_denovo/trinity_denovo_test/chrysalis/inchworm.DS.fa.min100 /path/to/trinity/working/directory/08_trinity_denovo/trinity_denovo_test/chrysalis/inchworm.DS.fa.min100
3872    /home/aperpos/programs/miniconda3/envs/trinity_venv/bin/bowtie2-build --wrapper basic-0 --threads 8 -o 3 /path/to/trinity/working/directory/08_trinity_denovo/trinity_denovo_test/chrysalis/inchworm.DS.fa.min100 /path/to/trinity/working/directory/08_trinity_denovo/trinity_denovo_test/chrysalis/inchworm.DS.fa.min100
```

When tracing system calls from htop, processes 3780 and 3781 appears as waiting on 3781 and 3782, respectively, and process 3782 is shown as:

```
strace: Process 3872 attached
restart_syscall(<... resuming interrupted nanosleep ...>) = 0
nanosleep({0, 1000000}, NULL)           = 0
nanosleep({0, 1000000}, NULL)           = 0
nanosleep({0, 1000000}, NULL)           = 0
nanosleep({0, 1000000}, NULL)           = 0
nanosleep({0, 1000000}, NULL)           = 0
nanosleep({0, 1000000}, NULL)           = 0
nanosleep({0, 1000000}, NULL)           = 0
nanosleep({0, 1000000}, NULL)           = 0
...
```
(note that I just copied a small excerpt)

Please note that, when I standalone-run the command that is put to sleep in my system, it works properly and all the files are successfully generated. Note I also managed to run Trinity from beginning to end in this same virtual environment using a different pair of FASTQ files.

This sounds very similar to the issue reported here. I am at a loss here since I do not know whether the issue is coming from the call from Trinity or from something inherent to bowtie2, as bowtie2 runs well outside of the trinity call. Hence why I am writing to both developers.

Do not hesitate to ask for any additional information if needed.

Thanks a lot,
Alberto

**Comment by ch4rr0** [comment]
Hello @apposada,

Can you try running with one thread and tail the output of the stderr file in order to determine whether the build process gets stuck?

**Comment by apposada** [comment]
Hi,

thanks for the reply. Since the bowtie2 step is embedded within the Trinity workflow, and since the bowtie2 command runs fine on its own, I am not sure how to do it but I will try. I will let you know whenever I have it, thanks.

**Comment by ch4rr0** [comment]
Would the `--CPU 1` option be the answer?

**Comment by dzolier** [comment]
I'm having the same problem. 
I'm using bowtie2-align-s version 2.4.2, and when I input the command (trying to do this with the wheat genome):
` bowtie2-build /WheatGenome/chr1A.fna,/WheatGenome/chr1B.fna,/WheatGenome/chr1D.fna,/WheatGenome/chr2A.fna,/WheatGenome/chr2B.fna,/WheatGenome/chr2D.fna,/WheatGenome/chr3A.fna,/WheatGenome/chr3B.fna,/WheatGenome/chr3D.fna,/WheatGenome/chr4A.fna,/WheatGenome/chr4B.fna,/WheatGenome/chr4D.fna,/WheatGenome/chr5A.fna,/WheatGenome/chr5B.fna,/WheatGenome/chr5D.fna,/WheatGenome/chr6A.fna,/WheatGenome/chr6B.fna,/WheatGenome/chr6D.fna,/WheatGenome/chr7A.fna,/WheatGenome/chr7B.fna,/WheatGenome/chr7D.fna wheat_ref`

I get this feedback:
`Settings:
  Output files: "wheat_ref.*.bt2l"
  Line rate: 7 (line is 128 bytes)
  Lines per side: 1 (side is 128 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr1A.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr1B.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr1D.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr2A.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr2B.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr2D.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr3A.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr3B.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr3D.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr4A.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr4B.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr4D.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr5A.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr5B.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr5D.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr6A.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr6B.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr6D.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr7A.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr7B.fna
  /mnt/d/Linux_subsystems/asugantya/WheatGenome/ncbi_dataset/data/GCA_018294505.1/chr7D.fna
Building a LARGE index
Reading reference sizes
  Time reading reference sizes: 00:04:38
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:03:44
bmax according to bmaxDivN setting: 3501751496
Using parameters --bmax 2626313622 --dcv 1024
  Doing ahead-of-time memory usage test`

The resulting files are wheat_ref.1.bt2l (4,304 KB), wheat_ref.2.bt2l (0 KB), wheat_ref.3.bt2l (3,049 KB), & wheat_ref.4.bt2l (3,419,680 KB)

It seems like my issue might involve a lack of RAM, since I'm doing this on a laptop with 16G, but it did the exact same thing as several other people on here have talked about, so I thought you guys might have more insight on the topic.

**Comment by drhoads** [comment]
I had the same problem using v2.3.4 in a docker container.  this has worked in the past multiple times to index chicken genome assemblies.  This time it did not, but then I found out that when Docker environment had updated the resources had been set back to the default, so when I upped the processors from 2 to 8, and RAM from 2 G to 80G, now it takes bowtie2-build a LOT longer to complete and I have the rev files.

**Comment by YunyunGao374** [comment]
@YiJessePi Hello, have you ever built the index successfully? I have the same question, that my reference data is about 90 Gb, and I have already run three and half days, and it still creates new files and data, that I never saw before with small data. Could you pls tell me how much time it cost for your data?  here are the temp files that I get.

23 Jun 12 10:29 rice.err
294263 Jun 15 15:58 rice.out
31784578950 Jun 14 17:40 rice_pangenome.1.bt2l.tmp
47648150644 Jun 14 17:40 rice_pangenome.2.bt2l.tmp
2422189 Jun 12 10:39 rice_pangenome.3.bt2l.tmp
23824075317 Jun 12 10:39 rice_pangenome.4.bt2l.tmp
96255674181 Jun 12 10:11 rice_pangenome.fasta
1611015032 Jun 15 15:42 rice_pangenome.rev.154.sa
934833304 Jun 15 15:52 rice_pangenome.rev.161.sa
964858928 Jun 15 15:43 rice_pangenome.rev.163.sa
11676422736 Jun 15 15:50 rice_pangenome.rev.1.bt2l.tmp
17507794948 Jun 15 15:50 rice_pangenome.rev.2.bt2l.tmp

**Comment by ch4rr0** [comment]
Hello,

Would you be willing to share your genome data? We have a server that should be able to build this genome. That way I can confirm whether it is a resource issue or some bug in the build code.

**Comment by ch4rr0** [comment]
Hello,

Would you be willing to share your genome data? We have a server that should be able to build this genome. That way I can confirm whether it is a resource issue or some bug in the build code.

**Comment by YunyunGao374** [comment]
@ch4rr0 Morning,

I just checked my result and found it has created the index! The reference genome is really big, which is about 90Gb, and it cost  about five days to finish it. Thanks, guys!

**Comment by rando2** [comment]
Hello! I'm having the same issue as other users, where bowtie2-build is not creating the rev indices. I am working on the [cat genome (felCat9)](https://api.ncbi.nlm.nih.gov/datasets/v2alpha/genome/accession/GCF_000181335.3/download?include_annotation_type=GENOME_FASTA). I started with the official release, then filtered it to include only the chromosomes (excluding any fragments that have not been assigned to a location). I'm getting the same issues even with the smaller fasta file.
